### PR TITLE
[PhpUnitBridge] add a triggered errors assertion helper

### PR DESCRIPTION
**Pull request by xabbuh**
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets |  |
| License | MIT |
| Doc PR |  |

In the past, we updated the test code several times that made sure that deprecation messages are triggered. I think we should move this code to a reusable class in the PhpUnitBridge to make it available everywhere and to be able to change a single method in case we need to update the logic.

### CODE REVIEW

**Comment by xabbuh** [comment]
@nicolas-grekas Any idea why the `deps=low` build is failing?

**Comment by Ener-Getick** [comment]
@xabbuh the phpunit-bridge is not required in several composer files such the one of the framework bundle (https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/FrameworkBundle/composer.json).

**Comment by Ener-Getick** [comment]
Anyway the tests probably won't pass as long as this isn't merged.

File: src/Symfony/Bridge/PhpUnit/ErrorAssert.php [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+namespace Symfony\Bridge\PhpUnit;
+
+/**
+ * Test that your code triggers expected error messages.
+ *
+ * @author Christian Flothmann <christian.flothmann@xabbuh.de>
+ */
+final class ErrorAssert
+{
+    public static function assertDeprecationsAreTriggered($expectedMessages, callable $testCode)
```
Comment (OskarStark) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: should we use singular here? `assertDeprecationIsTriggered` or anything like `setExpectedException` ? 

`setExpectedDeprecation` ?

Comment (stof) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: Given that the method takes an array of expected messages, plural is fine.

``setExpectedDeprecation`` is not fine for the current API. It would work if the method was register the expectation to check it later after running the test, but this is not what happens in the current API.

**Comment by stof** [comment]
@Ener-Getick we have a special phpunit launcher performing a local composer-based installation of PHPUnit (to remove unnecessary dependency which would hurt us, for instance ``symfony/yaml`` as we don't want to run tests on the version shipped in PHPUnit and we don't use the PHPUnit feature needing it). This special wrapper includes the bridge in the phpunit installation.

**Comment by stof** [comment]
@xabbuh looks like the PHPUnit setup installed the 3.1.0-beta1 version of the bridge, and so does not use the updated version of the bridge: https://travis-ci.org/symfony/symfony/jobs/133035973#L578

File: src/Symfony/Bridge/PhpUnit/ErrorAssert.php (hunk 1 above)
Comment (OskarStark) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: > Given that the method takes an array of expected messages, plural is fine.

Ok fine, but do we have many cases where we assert many deprecations?

> setExpectedDeprecation is not fine for the current API. It would work if the method was register the expectation to check it later after running the test, but this is not what happens in the current API.

👍

**Comment by Ener-Getick** [comment]
@stof thanks for the explanation, I didn't know that.

File: src/Symfony/Bridge/PhpUnit/ErrorAssert.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+            if ($expectedType !== $type) {
+                restore_error_handler();
+
+                return call_user_func_array('PHPUnit_Util_ErrorHandler::handleError', func_get_args());
+            }
+
+            $triggeredMessages[] = $message;
+        });
+
+        $testCode();
```
Comment (Ener-Getick) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: What about putting the entire function in a try/finally structure in case an exception is thrown ?
```php
try {
    // ...
} finally {
   restore_error_handler();
}

File: src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/FrameworkExtensionTest.php [hunk 3]
```diff
@@ -541,25 +542,12 @@ public function testSerializerCacheDisabled()
      */
     public function testDeprecatedSerializerCacheOption()
     {
-        $deprecations = array();
-        set_error_handler(function ($type, $msg) use (&$deprecations) {
-            if (E_USER_DEPRECATED !== $type) {
-                restore_error_handler();
+        ErrorAssert::assertDeprecationsAreTriggered('The "framework.serializer.cache" option is deprecated', function () {
```
Comment (Ener-Getick) [inline src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/FrameworkExtensionTest.php]: I'm not sure we should keep this changes now that 3.1 has been released, it may create merge conflicts in the future.

File: src/Symfony/Bridge/PhpUnit/ErrorAssert.php [hunk 4]
```diff
@@ -0,0 +1,41 @@
+<?php
+
```
Comment (Ener-Getick) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: The symfony header is missing

**Comment by fabpot** [comment]
ping @nicolas-grekas

File: src/Symfony/Bridge/PhpUnit/ErrorAssert.php [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    {
+        self::assertErrorsAreTriggered(E_USER_DEPRECATED, $expectedMessages, $testCode);
+    }
+
+    public static function assertErrorsAreTriggered($expectedType, $expectedMessages, callable $testCode)
+    {
+        $triggeredMessages = array();
+
+        set_error_handler(function ($type, $message) use ($expectedType, &$triggeredMessages) {
+            if ($expectedType !== $type) {
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: shouldn't this be a bitfield?

File: src/Symfony/Bridge/PhpUnit/ErrorAssert.php [hunk 6]
```diff
@@ -0,0 +1,41 @@
+<?php
+
+namespace Symfony\Bridge\PhpUnit;
+
+/**
+ * Test that your code triggers expected error messages.
+ *
+ * @author Christian Flothmann <christian.flothmann@xabbuh.de>
+ */
+final class ErrorAssert
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: what about making it a trait?

Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: what about making it a trait?

File: src/Symfony/Bridge/PhpUnit/ErrorAssert.php [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+        self::assertErrorsAreTriggered(E_USER_DEPRECATED, $expectedMessages, $testCode);
+    }
+
+    public static function assertErrorsAreTriggered($expectedType, $expectedMessages, callable $testCode)
+    {
+        $triggeredMessages = array();
+
+        set_error_handler(function ($type, $message) use ($expectedType, &$triggeredMessages) {
+            if ($expectedType !== $type) {
+                restore_error_handler();
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: why?

File: src/Symfony/Bridge/PhpUnit/ErrorAssert.php [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+                return call_user_func_array('PHPUnit_Util_ErrorHandler::handleError', func_get_args());
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: shouldn't this call the previous instead?

File: src/Symfony/Bridge/PhpUnit/ErrorAssert.php (hunk 7 above)
Comment (Ener-Getick) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: Because an exception is thrown and the function will stop here.
In case there are other tests the default handler must be restored.

Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: this could be better handled by your try/finally suggestion below, isn't it?

Comment (Ener-Getick) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: In fact as i said earlier we should use ``finally`` in case the phpunit handler ignores the error.

File: src/Symfony/Bridge/PhpUnit/ErrorAssert.php (hunk 8 above)
Comment (Ener-Getick) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: 👍  maybe with something like:
```php
set_error_handler($handler = function ($type, $message) use ($expectedType, &$triggeredMessages, &$handler) {
    // ...
    if () {
        restore_error_handler();
        trigger_error($message, $type);
        set_error_handler($handler);
   }
});

File: src/Symfony/Bridge/PhpUnit/ErrorAssert.php (hunk 7 above)
Comment (xabbuh) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: Are you sure that you can wrap the error handling code in a `try` block?

File: src/Symfony/Bridge/PhpUnit/ErrorAssert.php (hunk 6 above)
Comment (xabbuh) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: Not sure, what's the advantage given that the methods are static?

File: src/Symfony/Bridge/PhpUnit/ErrorAssert.php (hunk 5 above)
Comment (xabbuh) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: I don't think that that would be a good idea. I think if you really need to test for different kind of errors, you should probably rethink your code or write your own assertion code.

File: src/Symfony/Bridge/PhpUnit/ErrorAssert.php (hunk 6 above)
Comment (fabpot) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: indeed, static methods look fine to me.

File: src/Symfony/Bridge/PhpUnit/ErrorAssert.php (hunk 8 above)
Comment (xabbuh) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: done

File: src/Symfony/Bridge/PhpUnit/ErrorAssert.php (hunk 2 above)
Comment (xabbuh) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: done

File: src/Symfony/Bridge/PhpUnit/ErrorAssert.php (hunk 7 above)
Comment (Ener-Getick) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: @xabbuh what would be the problem ?

File: src/Symfony/Bridge/PhpUnit/ErrorAssert.php [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    public static function assertDeprecationsAreTriggered($expectedMessages, callable $testCode)
+    {
+        self::assertErrorsAreTriggered(E_USER_DEPRECATED, $expectedMessages, $testCode);
+    }
+
+    public static function assertErrorsAreTriggered($expectedType, $expectedMessages, callable $testCode)
+    {
+        $triggeredMessages = array();
+
+        try {
```
Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: what about this?
```php
        try {
            $prevHander = set_error_handler(function ($type, $msg, $file, $line, $context) use ($expectedType, &$triggeredMessages, &$prevHander) {
                if ($expectedType !== $type) {
                    return null !== $prevHandler && $prevHandler($type, $msg, $file, $line, $context);
                }
                $triggeredMessages[] = $msg;
            });
            $testCode();
        } finally {
            restore_error_handler();
        }
```

File: src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/FrameworkExtensionTest.php (hunk 3 above)
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/FrameworkExtensionTest.php]: In fact, these changes could be ported to 2.7 up to 3.1 (we did so when changing the bridge to add the clock mock previously)

File: src/Symfony/Bridge/PhpUnit/ErrorAssert.php (hunk 9 above)
Comment (Ener-Getick) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: :+1: looks clearer

Comment (xabbuh) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: That won't work if the previous handler was a class method. Though using `call_user_func()` might work:

```php
try {
    $prevHander = set_error_handler(function ($type, $msg, $file, $line, $context) use ($expectedType, &$triggeredMessages, &$prevHander) {
        if ($expectedType !== $type) {
            return null !== $prevHandler && call_user_func($prevHandler, $type, $msg, $file, $line, $context);
        }
        $triggeredMessages[] = $msg;
    });
    $testCode();
} finally {
    restore_error_handler();
}
```

What do you think?

File: src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/FrameworkExtensionTest.php (hunk 3 above)
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/FrameworkExtensionTest.php]: I would send another PR when this is merged for the `2.7` branch. Or would you prefer if we reverted the changes to the tests before merging and starting with them on `2.7`?

File: src/Symfony/Bridge/PhpUnit/ErrorAssert.php (hunk 9 above)
Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: ok for me (be careful, I made a typo handLLLLLer) :)

File: src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/FrameworkExtensionTest.php (hunk 3 above)
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/FrameworkExtensionTest.php]: Revert here and apply on 2.7 directly. That's how I did previously at least.

Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/FrameworkExtensionTest.php]: Updated here to contain only the new class. Tests will be updated after merging in separate PRs.

**Comment by nicolas-grekas** [comment]
:+1:

File: src/Symfony/Bridge/PhpUnit/ErrorAssert.php [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+            $prevHandler = set_error_handler(function ($type, $message, $file, $line, $context) use ($expectedType, &$triggeredMessages, &$prevHandler) {
+                if ($expectedType !== $type) {
+                    return null !== $prevHandler && call_user_func($prevHandler, $type, $message, $file, $line, $context);
+                }
+                $triggeredMessages[] = $message;
+            });
+
+            $testCode();
```
Comment (Ener-Getick) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: Isn't it safer to use ``call_user_func`` here as well ?

Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: Nope: no code has been written that can't be run. Having it here since the beginning will enforce this forever.

Comment (Ener-Getick) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: so we will only accept closure ?

Comment (Ener-Getick) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: so we won't accept other callables than closure ?

File: src/Symfony/Bridge/PhpUnit/ErrorAssert.php [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+ * @author Christian Flothmann <christian.flothmann@xabbuh.de>
+ */
+final class ErrorAssert
+{
+    public static function assertDeprecationsAreTriggered($expectedMessages, callable $testCode)
+    {
+        self::assertErrorsAreTriggered(E_USER_DEPRECATED, $expectedMessages, $testCode);
+    }
+
+    public static function assertErrorsAreTriggered($expectedType, $expectedMessages, callable $testCode)
```
Comment (stof) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: if we want to keep the bridge compatible with PHP 5.3, to use the latest version in 2.x builds, we need to remove the callable typehint

Comment (Ener-Getick) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: [it looks like the ``closure`` typehint works in php 5.3.](http://sandbox.onlinephpfunctions.com/code/a5bbbf91ff3020b19284c839fd323eb3355b9725)

File: src/Symfony/Bridge/PhpUnit/ErrorAssert.php (hunk 10 above)
Comment (xabbuh) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: @Ener-Getick But that would mean to drop support for other callables that are not anonymous functions.

Comment (Ener-Getick) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: @xabbuh indeed sorry, I thought the syntax ``$bar()`` only worked with closure...

File: src/Symfony/Bridge/PhpUnit/ErrorAssert.php (hunk 11 above)
Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: @stof good catch. What about allowing only closures here, and give $this as first arg when calling it (to work reduce the $that=$this boilerplate in 5.3?)

Comment (xabbuh) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: good catch, I removed the type hint and now do the check at runtime (throwing an exception if necessary)

File: src/Symfony/Bridge/PhpUnit/ErrorAssert.php [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+ */
+final class ErrorAssert
+{
+    /**
+     * @param string[] $expectedMessages Expected deprecation messages
+     * @param callable $testCode         A callable that is expected to trigger the expected deprecation messages when being executed
+     */
+    public static function assertDeprecationsAreTriggered($expectedMessages, $testCode)
+    {
+        if (!is_callable($testCode)) {
```
Comment (Ener-Getick) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: useless here as will be checked in the second function anyway

File: src/Symfony/Bridge/PhpUnit/ErrorAssert.php (hunk 10 above)
Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: I suggest doing `$testCode($this)` to ease writing tests in 5.3

File: src/Symfony/Bridge/PhpUnit/ErrorAssert.php (hunk 12 above)
Comment (xabbuh) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: That's true but I'd like to keep it here too for two reasons: Do not confuse users by letting them deal with exceptions that have been thrown by methods they did not call if we cannot avoid that. Secondly, mitigate the risk of breaking anything if we refactor the method that we call here.

File: src/Symfony/Bridge/PhpUnit/ErrorAssert.php (hunk 10 above)
Comment (Ener-Getick) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: @nicolas-grekas the class is static

File: src/Symfony/Bridge/PhpUnit/ErrorAssert.php (hunk 12 above)
Comment (Ener-Getick) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: ok fine for me :)

File: src/Symfony/Bridge/PhpUnit/ErrorAssert.php (hunk 10 above)
Comment (nicolas-grekas) [inline src/Symfony/Bridge/PhpUnit/ErrorAssert.php]: I forgot it is static :)

**Comment by nicolas-grekas** [comment]
Status: reviewed

**Comment by fabpot** [comment]
Thank you @xabbuh.
